ethpm/tools/builder.py

```diff
@@ -438,12 +441,222 @@ def validate_link_ref(offset: int, length: int, bytecode: str) -> str:
+                    raise ValidationError(f"Provided block: {block} is not valid hex.")
+
+                if tx_receipt.blockHash != block:
+                    raise ValidationError(
+                        f"Provided block: {block} does not match block hash found on "
+                        f"tx receipt: {tx_receipt.blockHash}."
+                    )
+    return _deployment(
+        contract_instance,
+        contract_type,
+        deployment_bytecode,
+        runtime_bytecode,
+        compiler,
+        uri,
+        address,
+        transaction,
+        block,
+        w3,
+    )
+
+
+def deployment(
+    *,
+    block_uri: URI,
+    contract_instance: str,
+    contract_type: str,
+    address: HexStr,
+    transaction: HexStr = None,
+    block: HexStr = None,
+    deployment_bytecode: Dict[str, Any] = None,
+    runtime_bytecode: Dict[str, Any] = None,
+    compiler: Dict[str, Any] = None,
+) -> Manifest:
+    validate_address(address)
+    return _deployment(
+        contract_instance,
+        contract_type,
+        deployment_bytecode,
+        runtime_bytecode,
+        compiler,
+        block_uri,
+        address,
+        transaction,
+        block,
+        None,
+    )
+
+
+@curry
+def _deployment(
+    contract_instance: str,
+    contract_type: str,
+    deployment_bytecode: Dict[str, Any],
+    runtime_bytecode: Dict[str, Any],
+    compiler: Dict[str, Any],
+    uri: URI,
+    address: HexStr,
+    tx: HexStr,
+    block: HexStr,
+    w3: Web3,
+    manifest: Manifest,
+) -> Manifest:
+    # todo: validate db,rb,compiler format
+    deployment_data = _build_dep_data(
+        contract_type,
+        deployment_bytecode,
+        runtime_bytecode,
+        compiler,
+        address,
+        tx,
+        block,
+        manifest,
+    )
+    if uri and not w3:
+        latest_block_uri = uri
+    elif w3:
+        latest_block_uri = create_block_uri_from_w3(w3)
+    else:
+        raise ManifestBuildingError(
+            "Must provide a valid blockchain URI or web3 instance."
+        )
+    return insert_deployment(
+        manifest, latest_block_uri, contract_instance, deployment_data
+    )
+
+
+@to_dict
+def _build_dep_data(
+    contract_type: str,
+    deployment_bytecode: Dict[str, Any],
+    runtime_bytecode: Dict[str, Any],
+    compiler: Dict[str, Any],
+    address: HexStr,
+    tx: HexStr,
+    block: HexStr,
+    manifest: Dict[str, Any],
+) -> Iterable[Tuple[str, Any]]:
+    yield "contract_type", contract_type
+    yield "address", to_hex(address)
+    if deployment_bytecode:
+        yield "deployment_bytecode", deployment_bytecode
+    if compiler:
+        yield "compiler", compiler
+    if tx:
+        yield "transaction", tx
+    if block:
+        yield "block", block
+    if runtime_bytecode:
+        if cytoolz.get_in(
+            ["contract_types", contract_type, "deployment_bytecode", "link_references"],
+            manifest,
+            0,
+        ):
```

Comment:
unsure how I feel about auto-populating the linked references from the deployment's corresponding contract type - another option is to simply force the user to provide the linked refs?

Reply:
If you are unsure about it I would go with the manual solution.  It's harder to remove functionality than it is to add it.